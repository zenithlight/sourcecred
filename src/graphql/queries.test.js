// @flow

import type {Body} from "./queries";
import {build, stringify, multilineLayout, inlineLayout} from "./queries";

describe("graphql/queries", () => {
  describe("end-to-end-test cases", () => {
    const testCases = {
      "a query using lots of features": featurefulQuery,
      "a useful query": usefulQuery,
    };
    Object.keys(testCases).forEach((key) => {
      describe(`for ${key}`, () => {
        const testCase = testCases[key];
        it("should build the query", () => {
          expect(testCase()).toMatchSnapshot();
        });
        it("should stringify as multiline", () => {
          const result = stringify.body(testCase(), multilineLayout("  "));
          expect(result).toMatchSnapshot();
        });
        it("should stringify as inline", () => {
          const result = stringify.body(testCase(), inlineLayout());
          expect(result).toMatchSnapshot();
        });
      });
    });
  });
  describe("complex object stringification", () => {
    const object = () => {
      const b = build;
      return b.object({
        exampleNumber: b.literal(123),
        exampleString: b.literal("hello"),
        exampleBoolean: b.literal(true),
        exampleNull: b.literal(null),
        exampleEnum: b.enumLiteral("WORLD"),
        exampleVariable: b.variable("widget"),
        exampleList: b.list([b.literal(12), b.literal(14), b.literal(16)]),
        exampleObject: b.object({
          roses: b.literal("red"),
          violets: b.literal("blue"),
        }),
        nestedList: b.list([
          b.list([b.literal("w"), b.literal("x")]),
          b.list([b.literal("y"), b.literal("z")]),
        ]),
        nestedObject: b.object({
          english: b.object({
            n1: b.literal("one"),
            n2: b.literal("two"),
          }),
          greek: b.object({
            n1: b.literal("ένα"),
            n2: b.literal("δύο"),
          }),
        }),
      });
    };
    it("should build the object", () => {
      expect(object()).toMatchSnapshot();
    });
    it("should work multiline", () => {
      const result = stringify.value(object(), multilineLayout("  "));
      expect(result).toMatchSnapshot();
    });
    it("should work inline", () => {
      const result = stringify.value(object(), inlineLayout());
      expect(result).toMatchSnapshot();
    });
  });
});

function featurefulQuery(): Body {
  const b = build;
  const body: Body = [
    b.query(
      "QueryWithParameters",
      [b.param("qp1", "String!"), b.param("qp2", "String!")],
      [
        b.field("thing", {id: b.literal(12345), name: b.variable("qp1")}, [
          b.field("fruit", {
            type: b.enumLiteral("APPLE"),
            tasty: b.literal(true),
          }),
          b.fragmentSpread("otherThings"),
        ]),
        b.field("more", {}, [
          b.inlineFragment("Widget", [
            b.field("mcguffins", {}, [b.field("quantity")]),
            b.alias(
              "goo",
              b.field("slime", {state: b.enumLiteral("SLIMY")}, [
                b.field("availability"),
              ])
            ),
          ]),
          b.inlineFragment("Gizmo", [
            b.field("cogs", {
              attributes: b.object({
                teeth: b.list([b.literal(12), b.literal(14), b.literal(16)]),
                shaft: b.literal(null),
              }),
            }),
          ]),
          b.inlineFragment(null, [b.field("__typename"), b.field("id")]),
        ]),
      ]
    ),
    b.query(
      "QueryWithoutParameters",
      [],
      [b.field("rateLimit", {}, [b.field("remaining")])]
    ),
    b.fragment("otherThings", "Thing", [
      b.field("__typename"),
      b.field("quality"),
    ]),
  ];
  return body;
}

function usefulQuery(): Body {
  const b = build;
  const makePageInfo = () => b.field("pageInfo", {}, [b.field("hasNextPage")]);
  const makeAuthor = () => b.field("author", {}, [b.fragmentSpread("whoami")]);
  const body: Body = [
    b.query(
      "FetchData",
      [b.param("owner", "String!"), b.param("name", "String!")],
      [
        b.field(
          "repository",
          {owner: b.variable("owner"), name: b.variable("name")},
          [
            b.field("issues", {first: b.literal(100)}, [
              makePageInfo(),
              b.field("nodes", {}, [
                b.field("id"),
                b.field("title"),
                b.field("body"),
                b.field("number"),
                makeAuthor(),
                b.field("comments", {first: b.literal(20)}, [
                  makePageInfo(),
                  b.field("nodes", {}, [
                    b.field("id"),
                    makeAuthor(),
                    b.field("body"),
                    b.field("url"),
                  ]),
                ]),
              ]),
            ]),
            b.field("pullRequests", {first: b.literal(100)}, [
              makePageInfo(),
              b.field("nodes", {}, [
                b.field("id"),
                b.field("title"),
                b.field("body"),
                b.field("number"),
                makeAuthor(),
                b.field("comments", {first: b.literal(20)}, [
                  makePageInfo(),
                  b.field("nodes", {}, [
                    b.field("id"),
                    makeAuthor(),
                    b.field("body"),
                    b.field("url"),
                  ]),
                ]),
                b.field("reviews", {first: b.literal(10)}, [
                  makePageInfo(),
                  b.field("nodes", {}, [
                    b.field("id"),
                    b.field("body"),
                    makeAuthor(),
                    b.field("state"),
                    b.field("comments", {first: b.literal(10)}, [
                      makePageInfo(),
                      b.field("nodes", {}, [
                        b.field("id"),
                        b.field("body"),
                        makeAuthor(),
                      ]),
                    ]),
                  ]),
                ]),
              ]),
            ]),
          ]
        ),
      ]
    ),
    b.fragment("whoami", "Actor", [
      b.field("__typename"),
      b.field("login"),
      b.inlineFragment("User", [b.field("id")]),
      b.inlineFragment("Organization", [b.field("id")]),
      b.inlineFragment("Bot", [b.field("id")]),
    ]),
  ];
  return body;
}
