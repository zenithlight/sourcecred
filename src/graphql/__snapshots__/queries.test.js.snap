// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql/queries complex object stringification should build the object 1`] = `
Object {
  "data": Object {
    "exampleBoolean": Object {
      "data": true,
      "type": "LITERAL",
    },
    "exampleEnum": Object {
      "data": "WORLD",
      "type": "ENUM",
    },
    "exampleList": Object {
      "data": Array [
        Object {
          "data": 12,
          "type": "LITERAL",
        },
        Object {
          "data": 14,
          "type": "LITERAL",
        },
        Object {
          "data": 16,
          "type": "LITERAL",
        },
      ],
      "type": "LIST",
    },
    "exampleNull": Object {
      "data": null,
      "type": "LITERAL",
    },
    "exampleNumber": Object {
      "data": 123,
      "type": "LITERAL",
    },
    "exampleObject": Object {
      "data": Object {
        "roses": Object {
          "data": "red",
          "type": "LITERAL",
        },
        "violets": Object {
          "data": "blue",
          "type": "LITERAL",
        },
      },
      "type": "OBJECT",
    },
    "exampleString": Object {
      "data": "hello",
      "type": "LITERAL",
    },
    "exampleVariable": Object {
      "data": "widget",
      "type": "VARIABLE",
    },
    "nestedList": Object {
      "data": Array [
        Object {
          "data": Array [
            Object {
              "data": "w",
              "type": "LITERAL",
            },
            Object {
              "data": "x",
              "type": "LITERAL",
            },
          ],
          "type": "LIST",
        },
        Object {
          "data": Array [
            Object {
              "data": "y",
              "type": "LITERAL",
            },
            Object {
              "data": "z",
              "type": "LITERAL",
            },
          ],
          "type": "LIST",
        },
      ],
      "type": "LIST",
    },
    "nestedObject": Object {
      "data": Object {
        "english": Object {
          "data": Object {
            "n1": Object {
              "data": "one",
              "type": "LITERAL",
            },
            "n2": Object {
              "data": "two",
              "type": "LITERAL",
            },
          },
          "type": "OBJECT",
        },
        "greek": Object {
          "data": Object {
            "n1": Object {
              "data": "ένα",
              "type": "LITERAL",
            },
            "n2": Object {
              "data": "δύο",
              "type": "LITERAL",
            },
          },
          "type": "OBJECT",
        },
      },
      "type": "OBJECT",
    },
  },
  "type": "OBJECT",
}
`;

exports[`graphql/queries complex object stringification should work inline 1`] = `"{ exampleNumber: 123 exampleString: \\"hello\\" exampleBoolean: true exampleNull: null exampleEnum: WORLD exampleVariable: $widget exampleList: [ 12 14 16 ] exampleObject: { roses: \\"red\\" violets: \\"blue\\" } nestedList: [ [ \\"w\\" \\"x\\" ] [ \\"y\\" \\"z\\" ] ] nestedObject: { english: { n1: \\"one\\" n2: \\"two\\" } greek: { n1: \\"ένα\\" n2: \\"δύο\\" } } }"`;

exports[`graphql/queries complex object stringification should work multiline 1`] = `
"{
  exampleNumber:
    123
  exampleString:
    \\"hello\\"
  exampleBoolean:
    true
  exampleNull:
    null
  exampleEnum:
    WORLD
  exampleVariable:
    $widget
  exampleList:
    [
      12
      14
      16
    ]
  exampleObject:
    {
      roses:
        \\"red\\"
      violets:
        \\"blue\\"
    }
  nestedList:
    [
      [
        \\"w\\"
        \\"x\\"
      ]
      [
        \\"y\\"
        \\"z\\"
      ]
    ]
  nestedObject:
    {
      english:
        {
          n1:
            \\"one\\"
          n2:
            \\"two\\"
        }
      greek:
        {
          n1:
            \\"ένα\\"
          n2:
            \\"δύο\\"
        }
    }
}"
`;

exports[`graphql/queries end-to-end-test cases for a query using lots of features should build the query 1`] = `
Array [
  Object {
    "name": "QueryWithParameters",
    "params": Array [
      Object {
        "name": "qp1",
        "type": "String!",
      },
      Object {
        "name": "qp2",
        "type": "String!",
      },
    ],
    "selections": Array [
      Object {
        "alias": null,
        "args": Object {
          "id": Object {
            "data": 12345,
            "type": "LITERAL",
          },
          "name": Object {
            "data": "qp1",
            "type": "VARIABLE",
          },
        },
        "name": "thing",
        "selections": Array [
          Object {
            "alias": null,
            "args": Object {
              "tasty": Object {
                "data": true,
                "type": "LITERAL",
              },
              "type": Object {
                "data": "APPLE",
                "type": "ENUM",
              },
            },
            "name": "fruit",
            "selections": Array [],
            "type": "FIELD",
          },
          Object {
            "fragmentName": "otherThings",
            "type": "FRAGMENT_SPREAD",
          },
        ],
        "type": "FIELD",
      },
      Object {
        "alias": null,
        "args": Object {},
        "name": "more",
        "selections": Array [
          Object {
            "selections": Array [
              Object {
                "alias": null,
                "args": Object {},
                "name": "mcguffins",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "quantity",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                ],
                "type": "FIELD",
              },
              Object {
                "alias": "goo",
                "args": Object {
                  "state": Object {
                    "data": "SLIMY",
                    "type": "ENUM",
                  },
                },
                "name": "slime",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "availability",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                ],
                "type": "FIELD",
              },
            ],
            "type": "INLINE_FRAGMENT",
            "typeCondition": "Widget",
          },
          Object {
            "selections": Array [
              Object {
                "alias": null,
                "args": Object {
                  "attributes": Object {
                    "data": Object {
                      "shaft": Object {
                        "data": null,
                        "type": "LITERAL",
                      },
                      "teeth": Object {
                        "data": Array [
                          Object {
                            "data": 12,
                            "type": "LITERAL",
                          },
                          Object {
                            "data": 14,
                            "type": "LITERAL",
                          },
                          Object {
                            "data": 16,
                            "type": "LITERAL",
                          },
                        ],
                        "type": "LIST",
                      },
                    },
                    "type": "OBJECT",
                  },
                },
                "name": "cogs",
                "selections": Array [],
                "type": "FIELD",
              },
            ],
            "type": "INLINE_FRAGMENT",
            "typeCondition": "Gizmo",
          },
          Object {
            "selections": Array [
              Object {
                "alias": null,
                "args": Object {},
                "name": "__typename",
                "selections": Array [],
                "type": "FIELD",
              },
              Object {
                "alias": null,
                "args": Object {},
                "name": "id",
                "selections": Array [],
                "type": "FIELD",
              },
            ],
            "type": "INLINE_FRAGMENT",
            "typeCondition": null,
          },
        ],
        "type": "FIELD",
      },
    ],
    "type": "QUERY",
  },
  Object {
    "name": "QueryWithoutParameters",
    "params": Array [],
    "selections": Array [
      Object {
        "alias": null,
        "args": Object {},
        "name": "rateLimit",
        "selections": Array [
          Object {
            "alias": null,
            "args": Object {},
            "name": "remaining",
            "selections": Array [],
            "type": "FIELD",
          },
        ],
        "type": "FIELD",
      },
    ],
    "type": "QUERY",
  },
  Object {
    "name": "otherThings",
    "selections": Array [
      Object {
        "alias": null,
        "args": Object {},
        "name": "__typename",
        "selections": Array [],
        "type": "FIELD",
      },
      Object {
        "alias": null,
        "args": Object {},
        "name": "quality",
        "selections": Array [],
        "type": "FIELD",
      },
    ],
    "type": "FRAGMENT",
    "typeCondition": "Thing",
  },
]
`;

exports[`graphql/queries end-to-end-test cases for a query using lots of features should stringify as inline 1`] = `"query QueryWithParameters($qp1: String! $qp2: String!) { thing(id: 12345 name: $qp1) { fruit(type: APPLE tasty: true) ...otherThings } more { ... on Widget { mcguffins { quantity } goo: slime(state: SLIMY) { availability } } ... on Gizmo { cogs(attributes: { teeth: [ 12 14 16 ] shaft: null }) } ... { __typename id } } } query QueryWithoutParameters { rateLimit { remaining } } fragment otherThings on Thing { __typename quality }"`;

exports[`graphql/queries end-to-end-test cases for a query using lots of features should stringify as multiline 1`] = `
"query QueryWithParameters($qp1: String! $qp2: String!) {
  thing(id: 12345 name: $qp1) {
    fruit(type: APPLE tasty: true)
    ...otherThings
  }
  more {
    ... on Widget {
      mcguffins {
        quantity
      }
      goo: slime(state: SLIMY) {
        availability
      }
    }
    ... on Gizmo {
      cogs(attributes: { teeth: [ 12 14 16 ] shaft: null })
    }
    ... {
      __typename
      id
    }
  }
}
query QueryWithoutParameters {
  rateLimit {
    remaining
  }
}
fragment otherThings on Thing {
  __typename
  quality
}"
`;

exports[`graphql/queries end-to-end-test cases for a useful query should build the query 1`] = `
Array [
  Object {
    "name": "FetchData",
    "params": Array [
      Object {
        "name": "owner",
        "type": "String!",
      },
      Object {
        "name": "name",
        "type": "String!",
      },
    ],
    "selections": Array [
      Object {
        "alias": null,
        "args": Object {
          "name": Object {
            "data": "name",
            "type": "VARIABLE",
          },
          "owner": Object {
            "data": "owner",
            "type": "VARIABLE",
          },
        },
        "name": "repository",
        "selections": Array [
          Object {
            "alias": null,
            "args": Object {
              "first": Object {
                "data": 100,
                "type": "LITERAL",
              },
            },
            "name": "issues",
            "selections": Array [
              Object {
                "alias": null,
                "args": Object {},
                "name": "pageInfo",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "hasNextPage",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                ],
                "type": "FIELD",
              },
              Object {
                "alias": null,
                "args": Object {},
                "name": "nodes",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "id",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "title",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "body",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "number",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "author",
                    "selections": Array [
                      Object {
                        "fragmentName": "whoami",
                        "type": "FRAGMENT_SPREAD",
                      },
                    ],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {
                      "first": Object {
                        "data": 20,
                        "type": "LITERAL",
                      },
                    },
                    "name": "comments",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "args": Object {},
                        "name": "pageInfo",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "hasNextPage",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                        ],
                        "type": "FIELD",
                      },
                      Object {
                        "alias": null,
                        "args": Object {},
                        "name": "nodes",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "id",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "author",
                            "selections": Array [
                              Object {
                                "fragmentName": "whoami",
                                "type": "FRAGMENT_SPREAD",
                              },
                            ],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "body",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "url",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                        ],
                        "type": "FIELD",
                      },
                    ],
                    "type": "FIELD",
                  },
                ],
                "type": "FIELD",
              },
            ],
            "type": "FIELD",
          },
          Object {
            "alias": null,
            "args": Object {
              "first": Object {
                "data": 100,
                "type": "LITERAL",
              },
            },
            "name": "pullRequests",
            "selections": Array [
              Object {
                "alias": null,
                "args": Object {},
                "name": "pageInfo",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "hasNextPage",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                ],
                "type": "FIELD",
              },
              Object {
                "alias": null,
                "args": Object {},
                "name": "nodes",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "id",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "title",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "body",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "number",
                    "selections": Array [],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {},
                    "name": "author",
                    "selections": Array [
                      Object {
                        "fragmentName": "whoami",
                        "type": "FRAGMENT_SPREAD",
                      },
                    ],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {
                      "first": Object {
                        "data": 20,
                        "type": "LITERAL",
                      },
                    },
                    "name": "comments",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "args": Object {},
                        "name": "pageInfo",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "hasNextPage",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                        ],
                        "type": "FIELD",
                      },
                      Object {
                        "alias": null,
                        "args": Object {},
                        "name": "nodes",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "id",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "author",
                            "selections": Array [
                              Object {
                                "fragmentName": "whoami",
                                "type": "FRAGMENT_SPREAD",
                              },
                            ],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "body",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "url",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                        ],
                        "type": "FIELD",
                      },
                    ],
                    "type": "FIELD",
                  },
                  Object {
                    "alias": null,
                    "args": Object {
                      "first": Object {
                        "data": 10,
                        "type": "LITERAL",
                      },
                    },
                    "name": "reviews",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "args": Object {},
                        "name": "pageInfo",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "hasNextPage",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                        ],
                        "type": "FIELD",
                      },
                      Object {
                        "alias": null,
                        "args": Object {},
                        "name": "nodes",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "id",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "body",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "author",
                            "selections": Array [
                              Object {
                                "fragmentName": "whoami",
                                "type": "FRAGMENT_SPREAD",
                              },
                            ],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {},
                            "name": "state",
                            "selections": Array [],
                            "type": "FIELD",
                          },
                          Object {
                            "alias": null,
                            "args": Object {
                              "first": Object {
                                "data": 10,
                                "type": "LITERAL",
                              },
                            },
                            "name": "comments",
                            "selections": Array [
                              Object {
                                "alias": null,
                                "args": Object {},
                                "name": "pageInfo",
                                "selections": Array [
                                  Object {
                                    "alias": null,
                                    "args": Object {},
                                    "name": "hasNextPage",
                                    "selections": Array [],
                                    "type": "FIELD",
                                  },
                                ],
                                "type": "FIELD",
                              },
                              Object {
                                "alias": null,
                                "args": Object {},
                                "name": "nodes",
                                "selections": Array [
                                  Object {
                                    "alias": null,
                                    "args": Object {},
                                    "name": "id",
                                    "selections": Array [],
                                    "type": "FIELD",
                                  },
                                  Object {
                                    "alias": null,
                                    "args": Object {},
                                    "name": "body",
                                    "selections": Array [],
                                    "type": "FIELD",
                                  },
                                  Object {
                                    "alias": null,
                                    "args": Object {},
                                    "name": "author",
                                    "selections": Array [
                                      Object {
                                        "fragmentName": "whoami",
                                        "type": "FRAGMENT_SPREAD",
                                      },
                                    ],
                                    "type": "FIELD",
                                  },
                                ],
                                "type": "FIELD",
                              },
                            ],
                            "type": "FIELD",
                          },
                        ],
                        "type": "FIELD",
                      },
                    ],
                    "type": "FIELD",
                  },
                ],
                "type": "FIELD",
              },
            ],
            "type": "FIELD",
          },
        ],
        "type": "FIELD",
      },
    ],
    "type": "QUERY",
  },
  Object {
    "name": "whoami",
    "selections": Array [
      Object {
        "alias": null,
        "args": Object {},
        "name": "__typename",
        "selections": Array [],
        "type": "FIELD",
      },
      Object {
        "alias": null,
        "args": Object {},
        "name": "login",
        "selections": Array [],
        "type": "FIELD",
      },
      Object {
        "selections": Array [
          Object {
            "alias": null,
            "args": Object {},
            "name": "id",
            "selections": Array [],
            "type": "FIELD",
          },
        ],
        "type": "INLINE_FRAGMENT",
        "typeCondition": "User",
      },
      Object {
        "selections": Array [
          Object {
            "alias": null,
            "args": Object {},
            "name": "id",
            "selections": Array [],
            "type": "FIELD",
          },
        ],
        "type": "INLINE_FRAGMENT",
        "typeCondition": "Organization",
      },
      Object {
        "selections": Array [
          Object {
            "alias": null,
            "args": Object {},
            "name": "id",
            "selections": Array [],
            "type": "FIELD",
          },
        ],
        "type": "INLINE_FRAGMENT",
        "typeCondition": "Bot",
      },
    ],
    "type": "FRAGMENT",
    "typeCondition": "Actor",
  },
]
`;

exports[`graphql/queries end-to-end-test cases for a useful query should stringify as inline 1`] = `"query FetchData($owner: String! $name: String!) { repository(owner: $owner name: $name) { issues(first: 100) { pageInfo { hasNextPage } nodes { id title body number author { ...whoami } comments(first: 20) { pageInfo { hasNextPage } nodes { id author { ...whoami } body url } } } } pullRequests(first: 100) { pageInfo { hasNextPage } nodes { id title body number author { ...whoami } comments(first: 20) { pageInfo { hasNextPage } nodes { id author { ...whoami } body url } } reviews(first: 10) { pageInfo { hasNextPage } nodes { id body author { ...whoami } state comments(first: 10) { pageInfo { hasNextPage } nodes { id body author { ...whoami } } } } } } } } } fragment whoami on Actor { __typename login ... on User { id } ... on Organization { id } ... on Bot { id } }"`;

exports[`graphql/queries end-to-end-test cases for a useful query should stringify as multiline 1`] = `
"query FetchData($owner: String! $name: String!) {
  repository(owner: $owner name: $name) {
    issues(first: 100) {
      pageInfo {
        hasNextPage
      }
      nodes {
        id
        title
        body
        number
        author {
          ...whoami
        }
        comments(first: 20) {
          pageInfo {
            hasNextPage
          }
          nodes {
            id
            author {
              ...whoami
            }
            body
            url
          }
        }
      }
    }
    pullRequests(first: 100) {
      pageInfo {
        hasNextPage
      }
      nodes {
        id
        title
        body
        number
        author {
          ...whoami
        }
        comments(first: 20) {
          pageInfo {
            hasNextPage
          }
          nodes {
            id
            author {
              ...whoami
            }
            body
            url
          }
        }
        reviews(first: 10) {
          pageInfo {
            hasNextPage
          }
          nodes {
            id
            body
            author {
              ...whoami
            }
            state
            comments(first: 10) {
              pageInfo {
                hasNextPage
              }
              nodes {
                id
                body
                author {
                  ...whoami
                }
              }
            }
          }
        }
      }
    }
  }
}
fragment whoami on Actor {
  __typename
  login
  ... on User {
    id
  }
  ... on Organization {
    id
  }
  ... on Bot {
    id
  }
}"
`;
